File: src/types/round.ts
/**
 * TypeScript mirror of kuma_core/mame/activity/round.py Pydantic models.
 * Spec: notes/specs/2026-05-04-mame-activity-integration.md §2.3
 *
 * Keep in sync with:
 *   - kuma_core/mame/activity/round.py (Round, RoundStatus, RoundErrorInfo)
 *   - src/store/exportSlice.ts getWorkspaceSnapshot / restoreWorkspace
 */

import type { PlateMeta, ActivityRecord, MergedRow } from "./mame/activity"
import type { ClassifyRoundResult, RoundFileEntry } from "./mame/strategy"

export type RoundStatus =
  | "design"
  | "ordered"
  | "ngs_done"
  | "activity_linked"
  | "exported"
  | "combinatorial"
  | "closed"
  | "error"

export interface RoundErrorInfo {
  stage: "upload" | "merge" | "export" | "handoff"
  message: string
  occurred_at: string
}

/** A file one round produced, with the moment it was written. */
export interface RoundArtifact {
  path: string
  produced_at: string
}

/**
 * An advisory answer, kept with the inputs that produced it.
 *
 * The answer alone cannot be re-examined later: it is a statement about one
 * ordered set of round files at one moment. Storing the files and the time
 * alongside it is what lets a restored answer say what it was about, and
 * `input_signature` is what lets the screen tell whether the list in front of
 * the operator is still that same list.
 */
export interface RoundAdvisoryRecord {
  /** The sidecar response verbatim, decision or not_assessable. */
  result: ClassifyRoundResult
  /** The files it was computed from, with the round numbers they were sent as. */
  inputs: RoundFileEntry[]
  /** ISO timestamp of when the answer came back. */
  decided_at: string
  /**
   * Identity of `inputs` including their contents, from roundFilesSignature
   * (lib/round/roundArtifacts.ts): each entry carries the moment the app wrote
   * that file, so rebuilding a round over the same path no longer looks like
   * the list the answer was computed from.
   */
  input_signature: string
}

export interface Round {
  id: string
  n: number
  created_at: string
  status: RoundStatus
  error_info: RoundErrorInfo | null
  plate_meta: PlateMeta
  design: Record<string, unknown>
  genotype: Record<string, unknown>
  activity: { records: ActivityRecord[]; plate_meta: PlateMeta } | null
  merged_table: MergedRow[]
  /**
   * The EVOLVEpro input step 4.1 built for this round, which step 4.2 reads
   * back as one entry of the round series (lib/round/roundArtifacts.ts).
   *
   * Optional because rounds restored from a snapshot written before this field
   * existed simply do not have it. Absent reads as "this round produced
   * nothing", which is the same answer those projects give today, so no
   * snapshot schema bump is warranted.
   */
  evolvepro_input?: RoundArtifact | null
  /**
   * The last advisory answer computed while this round was active.
   *
   * Optional for the same reason as `evolvepro_input`: absent on rounds
   * restored from an older snapshot, and absent means no advisory has been run,
   * which is what those projects report today.
   */
  advisory?: RoundAdvisoryRecord | null
}

File: kuma_core/mame/activity/round.py
"""Round entity Pydantic models for MAME integration.

Spec: notes/specs/2026-05-04-mame-activity-integration.md §2.1, §2.2
"""

from datetime import datetime
from enum import Enum
from typing import Literal
from pydantic import BaseModel
from kuma_core.mame.activity.models import PlateMeta, ActivityTable, MergedRow


class RoundStatus(str, Enum):
    DESIGN = "design"
    ORDERED = "ordered"
    NGS_DONE = "ngs_done"
    ACTIVITY_LINKED = "activity_linked"
    EXPORTED = "exported"
    COMBINATORIAL = "combinatorial"
    CLOSED = "closed"
    ERROR = "error"


class RoundErrorInfo(BaseModel):
    stage: Literal["upload", "merge", "export", "handoff"]
    message: str
    occurred_at: datetime


class RoundArtifact(BaseModel):
    """A file one round produced, with the moment it was written."""

    path: str
    produced_at: datetime


class EvolveproInputArtifact(RoundArtifact):
    """The step 4.1 workbook, plus the wild-type replicates behind it.

    The exported activity column is each variant measurement over the mean WT
    of its cohort, so the same division applied to the WT rows states the assay
    spread on that scale.  Those rows are filtered out of the workbook, which
    holds one row per designed variant for EVOLVEpro, leaving nothing in the
    file for step 4.2 to read them back from.

    They ride on the artifact rather than beside it so that a rebuild replaces
    the path, the timestamp, and the replicates in one move.  Split across two
    fields, a rebuild could leave one round replicate list attached to another
    round file.

    Empty when the primary source carried no WT rows (the pre-normalized GC
    sheet), and on every round built before this field existed.  Empty reads as
    "none on record", which is what those rounds have.
    """

    wt_values: list[float] = []


class Round(BaseModel):
    id: str
    n: int
    created_at: datetime
    status: RoundStatus
    error_info: RoundErrorInfo | None = None
    plate_meta: PlateMeta
    design: dict
    genotype: dict
    activity: ActivityTable | None
    merged_table: list[MergedRow]
    # The EVOLVEpro input MAME step 4.1 built for this round.  Defaults to None
    # because rounds saved before the field existed do not carry it, and absent
    # means the round produced nothing, which is what those rounds report.
    evolvepro_input: EvolveproInputArtifact | None = None
    # The last advisory answer computed while this round was active, as the
    # strategy.classify_round response plus the files and the moment it came
    # from (see RoundAdvisoryRecord in src/types/round.ts).  Kept as a plain
    # dict like design and genotype above: the response is a two-shape union
    # owned by the handler, and no Python caller reads this field.
    advisory: dict | None = None
